### PR TITLE
Add a GRPC client/server to kando to run/check processes

### PR DESCRIPTION
**Pull request by tdmanv**
#### Server
```
./kando process server -h
server

Usage:
  kando process server [flags]

Flags:
  -h, --help   help for server

Global Flags:
  -a, --address string     The path of a unix socket of the process server (default "/tmp/kanister.sock")
  -v, --verbosity string   Log level (debug, info, warn, error, fatal, panic) (default "warning")
```
#### Client
```
./kando process client -h
Send commands to the process server

Usage:
  kando process client [command]

Available Commands:
  create      create
  list        list
  output      output

Flags:
  -j, --as-json   Display output as json
  -h, --help      help for client

Global Flags:
  -a, --address string     The path of a unix socket of the process server (default "/tmp/kanister.sock")
  -v, --verbosity string   Log level (debug, info, warn, error, fatal, panic) (default "warning")

Use "kando process client [command] --help" for more information about a command.
```

### Generate proto updates:
```
:/go/src/github.com/kanisterio/kanister# ./build/protoc.sh
protoc '--go_out=.' '--go_opt=paths=source_relative' '--go-grpc_out=.' '--go-grpc_opt=paths=source_relative' ./pkg/kanx/kanx.proto
```

## Pull request type

Please check the type of change your PR introduces:
- [x] :sunflower: Feature

## Test Plan
```
go test -v ./pkg/kanx
=== RUN   Test
{"File":"pkg/kanx/kanx_test.go","Function":"github.com/kanisterio/kanister/pkg/kanx.(*KanXSuite).TestError.func1","Line":168,"address":{"Name":"/tmp/KanXSuite.TestError1266107833/kanx.sock","Net":"unix"},"level":"info","msg":"Listening on socket","time":"2024-04-18T22:29:25.93761862-07:00"}
{"File":"pkg/kanx/kanx_grpc.pb.go","Function":"github.com/kanisterio/kanister/pkg/kanx._ProcessService_CreateProcesses_Handler","Line":190,"level":"info","msg":"pid:482191 state:PROCESS_STATE_RUNNING","stderr":"/tmp/KanXSuite.TestError1266107833/kando.2022764142.stderr","stdout":"/tmp/KanXSuite.TestError1266107833/kando.100297899.stdout","time":"2024-04-18T22:29:25.939380505-07:00"}
{"File":"runtime/asm_amd64.s","Function":"runtime.goexit","Line":1695,"level":"info","msg":"pid:482191 state:PROCESS_STATE_FAILED exitCode:-1 exitErr:\"signal: killed\"","time":"2024-04-18T22:29:25.940065676-07:00"}
{"File":"runtime/asm_amd64.s","Function":"runtime.goexit","Line":1695,"level":"info","msg":"Gracefully stopping. Context canceled","time":"2024-04-18T22:29:25.941803878-07:00"}
{"File":"pkg/kanx/kanx_test.go","Function":"github.com/kanisterio/kanister/pkg/kanx.(*KanXSuite).TestLongProcess.func1","Line":115,"address":{"Name":"/tmp/KanXSuite.TestLongProcess3405729663/kanx.sock","Net":"unix"},"level":"info","msg":"Listening on socket","time":"2024-04-18T22:29:25.942031333-07:00"}
{"File":"pkg/kanx/kanx_grpc.pb.go","Function":"github.com/kanisterio/kanister/pkg/kanx._ProcessService_CreateProcesses_Handler","Line":190,"level":"info","msg":"pid:482195 state:PROCESS_STATE_RUNNING","stderr":"/tmp/KanXSuite.TestLongProcess3405729663/kando.640450080.stderr","stdout":"/tmp/KanXSuite.TestLongProcess3405729663/kando.670754590.stdout","time":"2024-04-18T22:29:25.942692496-07:00"}
{"File":"runtime/asm_amd64.s","Function":"runtime.goexit","Line":1695,"level":"info","msg":"pid:482195 state:PROCESS_STATE_FAILED exitCode:-1 exitErr:\"signal: killed\"","time":"2024-04-18T22:29:25.9431854-07:00"}
{"File":"runtime/asm_amd64.s","Function":"runtime.goexit","Line":1695,"level":"info","msg":"Gracefully stopping. Context canceled","time":"2024-04-18T22:29:25.946201522-07:00"}
{"File":"pkg/kanx/kanx_test.go","Function":"github.com/kanisterio/kanister/pkg/kanx.(*KanXSuite).TestParallelStdout.func1","Line":225,"address":{"Name":"/tmp/KanXSuite.TestParallelStdout3622194217/kanx.sock","Net":"unix"},"level":"info","msg":"Listening on socket","time":"2024-04-18T22:29:25.946449017-07:00"}
{"File":"pkg/kanx/kanx_grpc.pb.go","Function":"github.com/kanisterio/kanister/pkg/kanx._ProcessService_CreateProcesses_Handler","Line":190,"level":"info","msg":"pid:482196 state:PROCESS_STATE_RUNNING","stderr":"/tmp/KanXSuite.TestParallelStdout3622194217/kando.3676522233.stderr","stdout":"/tmp/KanXSuite.TestParallelStdout3622194217/kando.1459663735.stdout","time":"2024-04-18T22:29:25.947107068-07:00"}
{"File":"runtime/asm_amd64.s","Function":"runtime.goexit","Line":1695,"level":"info","msg":"pid:482196 state:PROCESS_STATE_FAILED exitCode:-1 exitErr:\"signal: killed\"","time":"2024-04-18T22:29:25.947828873-07:00"}
{"File":"runtime/asm_amd64.s","Function":"runtime.goexit","Line":1695,"level":"info","msg":"Gracefully stopping. Context canceled","time":"2024-04-18T22:29:26.082038582-07:00"}
{"File":"pkg/kanx/kanx_test.go","Function":"github.com/kanisterio/kanister/pkg/kanx.(*KanXSuite).TestServerCancellation","Line":67,"address":{"Name":"/tmp/KanXSuite.TestServerCancellation881028328/kanx.sock","Net":"unix"},"level":"info","msg":"Listening on socket","time":"2024-04-18T22:29:26.082330192-07:00"}
{"File":"runtime/asm_amd64.s","Function":"runtime.goexit","Line":1695,"level":"info","msg":"Gracefully stopping. Context canceled","time":"2024-04-18T22:29:26.082690116-07:00"}
{"File":"pkg/kanx/kanx_test.go","Function":"github.com/kanisterio/kanister/pkg/kanx.(*KanXSuite).TestShortProcess.func1","Line":77,"address":{"Name":"/tmp/KanXSuite.TestShortProcess1010242134/kanx.sock","Net":"unix"},"level":"info","msg":"Listening on socket","time":"2024-04-18T22:29:26.083038901-07:00"}
{"File":"pkg/kanx/kanx_grpc.pb.go","Function":"github.com/kanisterio/kanister/pkg/kanx._ProcessService_CreateProcesses_Handler","Line":190,"level":"info","msg":"pid:482209 state:PROCESS_STATE_RUNNING","stderr":"/tmp/KanXSuite.TestShortProcess1010242134/kando.1760524628.stderr","stdout":"/tmp/KanXSuite.TestShortProcess1010242134/kando.3646466058.stdout","time":"2024-04-18T22:29:26.161016192-07:00"}
{"File":"runtime/asm_amd64.s","Function":"runtime.goexit","Line":1695,"level":"info","msg":"pid:482209 state:PROCESS_STATE_SUCCEEDED","time":"2024-04-18T22:29:26.161705947-07:00"}
{"File":"runtime/asm_amd64.s","Function":"runtime.goexit","Line":1695,"level":"info","msg":"Gracefully stopping. Context canceled","time":"2024-04-18T22:29:26.163238424-07:00"}
OK: 5 passed
--- PASS: Test (0.23s)
PASS
ok      github.com/kanisterio/kanister/pkg/kanx 0.239s
```

- [x] :muscle: Manual
- [x] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

**Review by hairyhum — APPROVED** [review]

File: pkg/kanx/kanx_test.go [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+	err = Stderr(ctx, addr, p.GetPid(), buf)
+	c.Assert(err, IsNil)
+	c.Assert(buf.String(), Equals, "")
+}
+
+var _ io.Writer = (*nilWriter)(nil)
+
+type nilWriter struct{}
+
+func (nw *nilWriter) Write(p []byte) (n int, err error) { return }
```
Comment (denisvmedia) [inline pkg/kanx/kanx_test.go:218]: Why not to use io.Discard instead?

Comment (tdmanv) [reply pkg/kanx/kanx_test.go:218]: good suggestion. i hadn't seen io.Discard before

**Review by aaron-kasten — COMMENTED** [review]
I don't oppose using this tool, but I don't see how it adds to capturing log output (beyond what Kube `exec` already does).  The functionality appears to be a duplication of K8s `exec` with the addition of tracking output per-command.

Is there a use-case for storing and replaying output from kanister-tools pods?

File: pkg/kanx/kanx_test.go (hunk 1 above)
Comment (aaron-kasten) [reply pkg/kanx/kanx_test.go]: @denisvmedia `io.Discard` is used in the new changes

**Comment by hairyhum** [comment]
> The functionality appears to be a duplication of K8s exec with the addition of tracking output per-command.
I believe the main difference is that unlike `exec` this one will not break the long-running command execution if `exec` interrupts.

File: pkg/kanx/server.go [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+	}
+	return s.streamOutput(ss, p, fh)
+}
+
+type sender interface {
+	Send(*Output) error
+}
+
+func (s *processServiceServer) streamOutput(ss sender, p *process, fh *os.File) error {
+	buf := bytes.NewBuffer(make([]byte, 0, 4*1024*1024)) // 4MiB is the max size of a GRPC request
```
Comment (denisvmedia) [inline pkg/kanx/server.go:138]: I'd extract this magic number as a constant.

File: pkg/kanx/client.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	"net/url"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+func unixDialer(ctx context.Context, addr string) (net.Conn, error) {
+	return net.Dial("unix", addr)
+}
+func newGRPCConnection(addr string) (*grpc.ClientConn, error) {
```
Comment (denisvmedia) [inline pkg/kanx/client.go:16]: nit

```suggestion
}

func newGRPCConnection(addr string) (*grpc.ClientConn, error) {
```

File: pkg/kanx/client.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+func ListProcesses(ctx context.Context, addr string) ([]*Process, error) {
+	conn, err := newGRPCConnection(addr)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+	c := NewProcessServiceClient(conn)
+	in := &ListProcessesRequest{}
+	lpc, err := c.ListProcesses(ctx, in)
```
Comment (denisvmedia) [inline pkg/kanx/client.go:50]: redundant var?

```suggestion
	lpc, err := c.ListProcesses(ctx, &ListProcessesRequest{})
```

**Review by denisvmedia — APPROVED** [review]

**Pull request merged**